# Machine Learning With Spark ML
In this lab assignment, you will complete a project by going through the following steps:
1. Get the data.
2. Discover the data to gain insights.
3. Prepare the data for Machine Learning algorithms.
4. Select a model and train it.
5. Fine-tune your model.
6. Present your solution.

As a dataset, we use the California Housing Prices dataset from the StatLib repository. This dataset was based on data from the 1990 California census. The dataset has the following columns
1. `longitude`: a measure of how far west a house is (a higher value is farther west)
2. `latitude`: a measure of how far north a house is (a higher value is farther north)
3. `housing_,median_age`: median age of a house within a block (a lower number is a newer building)
4. `total_rooms`: total number of rooms within a block
5. `total_bedrooms`: total number of bedrooms within a block
6. `population`: total number of people residing within a block
7. `households`: total number of households, a group of people residing within a home unit, for a block
8. `median_income`: median income for households within a block of houses
9. `median_house_value`: median house value for households within a block
10. `ocean_proximity`: location of the house w.r.t ocean/sea

---
# 1. Get the data
Let's start the lab by loading the dataset. The can find the dataset at `data/housing.csv`. To infer column types automatically, when you are reading the file, you need to set `inferSchema` to true. Moreover enable the `header` option to read the columns' name from the file.

In [0]:
#Imports
from pyspark.sql.functions import *

In [0]:
# TODO: Replace <FILL IN> with appropriate code

#val housing = spark.read.<FILL IN>.load("data/housing.csv")
#%scala

housing = spark.read.format("csv").load("dbfs:/FileStore/shared_uploads/ravib@kth.se/housing.csv",header=True, inferSchema= True)

housing.show(5)

+---------+--------+------------------+-----------+--------------+----------+----------+-------------+------------------+---------------+
longitude|latitude|housing_median_age|total_rooms|total_bedrooms|population|households|median_income|median_house_value|ocean_proximity|
+---------+--------+------------------+-----------+--------------+----------+----------+-------------+------------------+---------------+
 -122.23| 37.88| 41.0| 880.0| 129.0| 322.0| 126.0| 8.3252| 452600.0| NEAR BAY|
 -122.22| 37.86| 21.0| 7099.0| 1106.0| 2401.0| 1138.0| 8.3014| 358500.0| NEAR BAY|
 -122.24| 37.85| 52.0| 1467.0| 190.0| 496.0| 177.0| 7.2574| 352100.0| NEAR BAY|
 -122.25| 37.85| 52.0| 1274.0| 235.0| 558.0| 219.0| 5.6431| 341300.0| NEAR BAY|
 -122.25| 37.85| 52.0| 1627.0| 280.0| 565.0| 259.0| 3.8462| 342200.0| NEAR BAY|
+---------+--------+------------------+-----------+--------------+----------+----------+-------------+------------------+---------------+
only showing top 5 rows

---
# 2. Discover the data to gain insights
Now it is time to take a look at the data. In this step we are going to take a look at the data a few different ways:
* See the schema and dimension of the dataset
* Look at the data itself
* Statistical summary of the attributes
* Breakdown of the data by the categorical attribute variable
* Find the correlation among different attributes
* Make new attributes by combining existing attributes

## 2.1. Schema and dimension
Print the schema of the dataset

In [0]:
#// TODO: Replace <FILL IN> with appropriate code

housing.printSchema()

root
-- longitude: double (nullable = true)
-- latitude: double (nullable = true)
-- housing_median_age: double (nullable = true)
-- total_rooms: double (nullable = true)
-- total_bedrooms: double (nullable = true)
-- population: double (nullable = true)
-- households: double (nullable = true)
-- median_income: double (nullable = true)
-- median_house_value: double (nullable = true)
-- ocean_proximity: string (nullable = true)

Print the number of records in the dataset.

In [0]:
#// TODO: Replace <FILL IN> with appropriate code

housing.count()

Out[12]: 20640

## 2.2. Look at the data
Print the first five records of the dataset.

In [0]:
#// TODO: Replace <FILL IN> with appropriate code

housing.show(5)

+---------+--------+------------------+-----------+--------------+----------+----------+-------------+------------------+---------------+
longitude|latitude|housing_median_age|total_rooms|total_bedrooms|population|households|median_income|median_house_value|ocean_proximity|
+---------+--------+------------------+-----------+--------------+----------+----------+-------------+------------------+---------------+
 -122.23| 37.88| 41.0| 880.0| 129.0| 322.0| 126.0| 8.3252| 452600.0| NEAR BAY|
 -122.22| 37.86| 21.0| 7099.0| 1106.0| 2401.0| 1138.0| 8.3014| 358500.0| NEAR BAY|
 -122.24| 37.85| 52.0| 1467.0| 190.0| 496.0| 177.0| 7.2574| 352100.0| NEAR BAY|
 -122.25| 37.85| 52.0| 1274.0| 235.0| 558.0| 219.0| 5.6431| 341300.0| NEAR BAY|
 -122.25| 37.85| 52.0| 1627.0| 280.0| 565.0| 259.0| 3.8462| 342200.0| NEAR BAY|
+---------+--------+------------------+-----------+--------------+----------+----------+-------------+------------------+---------------+
only showing top 5 rows

Print the number of records with population more than 10000.

In [0]:
#// TODO: Replace <FILL IN> with appropriate code

val = housing.filter("population>'10000.0'")
print(val.count())

23

## 2.3. Statistical summary
Print a summary of the table statistics for the attributes `housing_median_age`, `total_rooms`, `median_house_value`, and `population`. You can use the `describe` command.

In [0]:
#// TODO: Replace <FILL IN> with appropriate code

housing.describe("housing_median_age", "total_rooms", "median_house_value", "population")

Out[15]: DataFrame[summary: string, housing_median_age: string, total_rooms: string, median_house_value: string, population: string]

Print the maximum age (`housing_median_age`), the minimum number of rooms (`total_rooms`), and the average of house values (`median_house_value`).

In [0]:
#// TODO: Replace <FILL IN> with appropriate code


housing.agg({'housing_median_age': 'max'}).show()
housing.agg({'total_rooms': 'min'}).show()
housing.agg({'median_house_value': 'avg'}).show()

+-----------------------+
max(housing_median_age)|
+-----------------------+
 52.0|
+-----------------------+

+----------------+
min(total_rooms)|
+----------------+
 2.0|
+----------------+

+-----------------------+
avg(median_house_value)|
+-----------------------+
 206855.81690891474|
+-----------------------+

## 2.4. Breakdown the data by categorical data
Print the number of houses in different areas (`ocean_proximity`), and sort them in descending order.

In [0]:
#// TODO: Replace <FILL IN> with appropriate code


housing.groupBy('ocean_proximity').count().sort(col("count").desc()).show()
   


+---------------+-----+
ocean_proximity|count|
+---------------+-----+
 <1H OCEAN| 9136|
 INLAND| 6551|
 NEAR OCEAN| 2658|
 NEAR BAY| 2290|
 ISLAND| 5|
+---------------+-----+

Print the average value of the houses (`median_house_value`) in different areas (`ocean_proximity`), and call the new column `avg_value` when print it.

In [0]:
#// TODO: Replace <FILL IN> with appropriate code

housing.groupBy('ocean_proximity').agg(avg('median_house_value').alias('avg_value')).show()

+---------------+------------------+
ocean_proximity| avg_value|
+---------------+------------------+
 ISLAND| 380440.0|
 NEAR OCEAN|249433.97742663656|
 NEAR BAY|259212.31179039303|
 <1H OCEAN|240084.28546409807|
 INLAND|124805.39200122119|
+---------------+------------------+

Rewrite the above question in SQL.

In [0]:
#// TODO: Replace <FILL IN> with appropriate code

housing.createOrReplaceTempView("df")
spark.sql("""select ocean_proximity, avg(median_house_value) as avg_value
             from df
             group by ocean_proximity
             """).show()

+---------------+------------------+
ocean_proximity| avg_value|
+---------------+------------------+
 ISLAND| 380440.0|
 NEAR OCEAN|249433.97742663656|
 NEAR BAY|259212.31179039303|
 <1H OCEAN|240084.28546409807|
 INLAND|124805.39200122119|
+---------------+------------------+

## 2.5. Correlation among attributes
Print the correlation among the attributes `housing_median_age`, `total_rooms`, `median_house_value`, and `population`. To do so, first you need to put these attributes into one vector. Then, compute the standard correlation coefficient (Pearson) between every pair of attributes in this new vector. To make a vector of these attributes, you can use the `VectorAssembler` Transformer.

In [0]:
#// TODO: Replace <FILL IN> with appropriate code

from pyspark.ml.feature import VectorAssembler

assembler = VectorAssembler(
    inputCols=["housing_median_age", "total_rooms", "median_house_value", "population"],
    outputCol="housingAttrs")

output = assembler.transform(housing)
output.select("housingAttrs").show(5, truncate=False)

+-----------------------------+
housingAttrs |
+-----------------------------+
[41.0,880.0,452600.0,322.0] |
[21.0,7099.0,358500.0,2401.0]|
[52.0,1467.0,352100.0,496.0] |
[52.0,1274.0,341300.0,558.0] |
[52.0,1627.0,342200.0,565.0] |
+-----------------------------+
only showing top 5 rows

In [0]:
#// TODO: Replace <FILL IN> with appropriate code



from pyspark.ml.stat import Correlation

r1 = Correlation.corr(output, "housingAttrs").head()
print(r1)
print("Pearson correlation matrix:\n" + str(r1[0]))



Row(pearson(housingAttrs)=DenseMatrix(4, 4, [1.0, -0.3613, 0.1056, -0.2962, -0.3613, 1.0, 0.1342, 0.8571, 0.1056, 0.1342, 1.0, -0.0246, -0.2962, 0.8571, -0.0246, 1.0], False))
Pearson correlation matrix:
DenseMatrix([[ 1. , -0.3612622 , 0.10562341, -0.29624424],
 [-0.3612622 , 1. , 0.13415311, 0.85712597],
 [ 0.10562341, 0.13415311, 1. , -0.02464968],
 [-0.29624424, 0.85712597, -0.02464968, 1. ]])

## 2.6. Combine and make new attributes
Now, let's try out various attribute combinations. In the given dataset, the total number of rooms in a block is not very useful, if we don't know how many households there are. What we really want is the number of rooms per household. Similarly, the total number of bedrooms by itself is not very useful, and we want to compare it to the number of rooms. And the population per household seems like also an interesting attribute combination to look at. To do so, add the three new columns to the dataset as below. We will call the new dataset the `housingExtra`.
```
rooms_per_household = total_rooms / households
bedrooms_per_room = total_bedrooms / total_rooms
population_per_household = population / households
```

In [0]:
#// TODO: Replace <FILL IN> with appropriate code

housingCol1 = housing.withColumn("rooms_per_household", col('total_rooms')/col('households'))
housingCol2 = housingCol1.withColumn("bedrooms_per_room", col('total_bedrooms')/col('total_rooms'))
housingExtra = housingCol2.withColumn("population_per_household", col('population')/col('households'))

housingExtra.select("rooms_per_household", "bedrooms_per_room", "population_per_household").show(5)

+-------------------+-------------------+------------------------+
rooms_per_household| bedrooms_per_room|population_per_household|
+-------------------+-------------------+------------------------+
 6.984126984126984|0.14659090909090908| 2.5555555555555554|
 6.238137082601054|0.15579659106916466| 2.109841827768014|
 8.288135593220339|0.12951601908657123| 2.8022598870056497|
 5.8173515981735155|0.18445839874411302| 2.547945205479452|
 6.281853281853282| 0.1720958819913952| 2.1814671814671813|
+-------------------+-------------------+------------------------+
only showing top 5 rows

---
## 3. Prepare the data for Machine Learning algorithms
Before going through the Machine Learning steps, let's first rename the label column from `median_house_value` to `label`.

In [0]:
#// TODO: Replace <FILL IN> with appropriate code

renamedHousing = housingExtra.withColumnRenamed("median_house_value", "label")
renamedHousing.show(2)



+---------+--------+------------------+-----------+--------------+----------+----------+-------------+--------+---------------+-------------------+-------------------+------------------------+
longitude|latitude|housing_median_age|total_rooms|total_bedrooms|population|households|median_income| label|ocean_proximity|rooms_per_household| bedrooms_per_room|population_per_household|
+---------+--------+------------------+-----------+--------------+----------+----------+-------------+--------+---------------+-------------------+-------------------+------------------------+
 -122.23| 37.88| 41.0| 880.0| 129.0| 322.0| 126.0| 8.3252|452600.0| NEAR BAY| 6.984126984126984|0.14659090909090908| 2.5555555555555554|
 -122.22| 37.86| 21.0| 7099.0| 1106.0| 2401.0| 1138.0| 8.3014|358500.0| NEAR BAY| 6.238137082601054|0.15579659106916466| 2.109841827768014|
+---------+--------+------------------+-----------+--------------+----------+----------+-------------+--------+---------------+-------------------+-------------------+------------------------+
only showing top 2 rows

Now, we want to separate the numerical attributes from the categorical attribute (`ocean_proximity`) and keep their column names in two different lists. Moreover, sice we don't want to apply the same transformations to the predictors (features) and the label, we should remove the label attribute from the list of predictors.

In [0]:
#// label columns
#colLabel = "label"

#// categorical columns
colCat = "ocean_proximity"

#// numerical columns
colNum = renamedHousing.schema.names
colNum.remove('label')
colNum.remove('ocean_proximity')
colCat = [colCat]
print(colCat)
print(colNum)

['ocean_proximity']
['longitude', 'latitude', 'housing_median_age', 'total_rooms', 'total_bedrooms', 'population', 'households', 'median_income', 'rooms_per_household', 'bedrooms_per_room', 'population_per_household']

## 3.1. Prepare continuse attributes
### Data cleaning
Most Machine Learning algorithms cannot work with missing features, so we should take care of them. As a first step, let's find the columns with missing values in the numerical attributes. To do so, we can print the number of missing values of each continues attributes, listed in `colNum`.

In [0]:
#// TODO: Replace <FILL IN> with appropriate code


renamedHousing.select([count(when(col(c).isNull(), c)).alias(c) for c in renamedHousing.columns]).show()


+---------+--------+------------------+-----------+--------------+----------+----------+-------------+-----+---------------+-------------------+-----------------+------------------------+
longitude|latitude|housing_median_age|total_rooms|total_bedrooms|population|households|median_income|label|ocean_proximity|rooms_per_household|bedrooms_per_room|population_per_household|
+---------+--------+------------------+-----------+--------------+----------+----------+-------------+-----+---------------+-------------------+-----------------+------------------------+
 0| 0| 0| 0| 207| 0| 0| 0| 0| 0| 0| 207| 0|
+---------+--------+------------------+-----------+--------------+----------+----------+-------------+-----+---------------+-------------------+-----------------+------------------------+

As we observerd above, the `total_bedrooms` and `bedrooms_per_room` attributes have some missing values. One way to take care of missing values is to use the `Imputer` Transformer, which completes missing values in a dataset, either using the mean or the median of the columns in which the missing values are located. To use it, you need to create an `Imputer` instance, specifying that you want to replace each attribute's missing values with the "median" of that attribute.

In [0]:
#// TODO: Replace <FILL IN> with appropriate code

#import org.apache.spark.ml.feature.Imputer
from pyspark.ml.feature import Imputer



imputer = Imputer(
    inputCols=["total_bedrooms", "bedrooms_per_room"],
    outputCols=["total_bedrooms", "bedrooms_per_room"]   
    
)

imputedHousing = imputer.setStrategy("median").fit(renamedHousing).transform(renamedHousing)

imputedHousing.select("total_bedrooms", "bedrooms_per_room").show(5)
imputedHousing.select([count(when(col(c).isNull(), c)).alias(c) for c in imputedHousing.columns]).show()

+--------------+-------------------+
total_bedrooms| bedrooms_per_room|
+--------------+-------------------+
 129.0|0.14659090909090908|
 1106.0|0.15579659106916466|
 190.0|0.12951601908657123|
 235.0|0.18445839874411302|
 280.0| 0.1720958819913952|
+--------------+-------------------+
only showing top 5 rows

+---------+--------+------------------+-----------+--------------+----------+----------+-------------+-----+---------------+-------------------+-----------------+------------------------+
longitude|latitude|housing_median_age|total_rooms|total_bedrooms|population|households|median_income|label|ocean_proximity|rooms_per_household|bedrooms_per_room|population_per_household|
+---------+--------+------------------+-----------+--------------+----------+----------+-------------+-----+---------------+-------------------+-----------------+------------------------+
 0| 0| 0| 0| 0| 0| 0| 0| 0| 0| 0| 0| 0|
+---------+--------+------------------+-----------+--------------+----------+----------+-------------+-----+---------------+-------------------+-----------------+------------------------+

### Scaling
One of the most important transformations you need to apply to your data is feature scaling. With few exceptions, Machine Learning algorithms don't perform well when the input numerical attributes have very different scales. This is the case for the housing data: the total number of rooms ranges from about 6 to 39,320, while the median incomes only range from 0 to 15. Note that scaling the label attribues is generally not required.

One way to get all attributes to have the same scale is to use standardization. In standardization, for each value, first it subtracts the mean value (so standardized values always have a zero mean), and then it divides by the variance so that the resulting distribution has unit variance. To do this, we can use the `StandardScaler` Estimator. To use `StandardScaler`, again we need to convert all the numerical attributes into a big vectore of features using `VectorAssembler`, and then call `StandardScaler` on that vactor.

In [0]:
#// TODO: Replace <FILL IN> with appropriate code
from pyspark.ml.feature import VectorAssembler, StandardScaler

assembler = VectorAssembler(
    inputCols=["housing_median_age", "total_rooms", "total_bedrooms", "population","households","median_income","rooms_per_household", "bedrooms_per_room", "population_per_household"],
    outputCol="features")

featuredHousing = assembler.transform(imputedHousing)


scaler = StandardScaler(
  inputCol="features",
  outputCol="scaled_features",
  withStd = True,
  withMean = True)
 
scaledHousing = scaler.fit(featuredHousing).transform(featuredHousing)

scaledHousing.select("features", "scaled_features").show(5, truncate=False)


#val va = new VectorAssembler().<FILL IN>
#val featuredHousing = va.transform(imputedHousing)

#val scaler = new StandardScaler().<FILL IN>
#val scaledHousing = scaler.fit(featuredHousing).transform(featuredHousing)

#scaledHousing.show(5)

+-------------------------------------------------------------------------------------------------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
features |scaled_features |
+-------------------------------------------------------------------------------------------------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
[41.0,880.0,129.0,322.0,126.0,8.3252,6.984126984126984,0.14659090909090908,2.5555555555555554] |[0.9821188656747666,-0.804799599801809,-0.9724529206674336,-0.9744049915469923,-0.977009185045236,2.3447089561176147,0.6285442264151613,-1.149894671563979,-0.049595334664572825] |
[21.0,7099.0,1106.0,2401.0,1138.0,8.3014,6.238137082601054,0.15579659106916466,2.109841827768014]|[-0.6070042082805048,2.0458405373571247,1.3571105530973857,0.8614179998296625,1.6699205725917976,2.3321814648403056,0.32703343493534437,-0.990349761433348,-0.09250998908805363] |
[52.0,1467.0,190.0,496.0,177.0,7.2574,8.288135593220339,0.12951601908657123,2.8022598870056497] |[1.8561365563501657,-0.5357329073251578,-0.8270042288561193,-0.8207574684591915,-0.8436164798678757,1.782656217213844,1.155592470660819,-1.4458218823526268,-0.025841901757064786] |
[52.0,1274.0,235.0,558.0,219.0,5.6431,5.8173515981735155,0.18445839874411302,2.547945205479452] |[1.8561365563501657,-0.6241994689060669,-0.7197060135854777,-0.7660095004623889,-0.7337636638394612,0.9329449075937323,0.15696227955717026,-0.4936081250741483,-0.050328081993623826]|
[52.0,1627.0,280.0,565.0,259.0,3.8462,6.281853281853282,0.1720958819913952,2.1814671814671813] |[1.8561365563501657,-0.4623927526466838,-0.612407798314836,-0.7598282782692015,-0.6291419342885903,-0.012880683843038048,0.344702448698221,-0.7078645624526793,-0.08561368522679019] |
+-------------------------------------------------------------------------------------------------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
only showing top 5 rows

## 3.2. Prepare categorical attributes
After imputing and scaling the continuse attributes, we should take care of the categorical attributes. Let's first print the number of distict values of the categirical attribute `ocean_proximity`.

In [0]:
#// TODO: Replace <FILL IN> with appropriate code

renamedHousing.select("ocean_proximity").distinct().count()

Out[258]: 5

### String indexer
Most Machine Learning algorithms prefer to work with numbers. So let's convert the categorical attribute `ocean_proximity` to numbers. To do so, we can use the `StringIndexer` that encodes a string column of labels to a column of label indices. The indices are in [0, numLabels), ordered by label frequencies, so the most frequent label gets index 0.

In [0]:
#// TODO: Replace <FILL IN> with appropriate code

from pyspark.ml.feature import StringIndexer

indexer = StringIndexer(
  inputCol="ocean_proximity",
  outputCol="indx_ocean_proximity")
  
idxHousing = indexer.fit(renamedHousing).transform(renamedHousing)

idxHousing.show(5)

+---------+--------+------------------+-----------+--------------+----------+----------+-------------+--------+---------------+-------------------+-------------------+------------------------+--------------------+
longitude|latitude|housing_median_age|total_rooms|total_bedrooms|population|households|median_income| label|ocean_proximity|rooms_per_household| bedrooms_per_room|population_per_household|indx_ocean_proximity|
+---------+--------+------------------+-----------+--------------+----------+----------+-------------+--------+---------------+-------------------+-------------------+------------------------+--------------------+
 -122.23| 37.88| 41.0| 880.0| 129.0| 322.0| 126.0| 8.3252|452600.0| NEAR BAY| 6.984126984126984|0.14659090909090908| 2.5555555555555554| 3.0|
 -122.22| 37.86| 21.0| 7099.0| 1106.0| 2401.0| 1138.0| 8.3014|358500.0| NEAR BAY| 6.238137082601054|0.15579659106916466| 2.109841827768014| 3.0|
 -122.24| 37.85| 52.0| 1467.0| 190.0| 496.0| 177.0| 7.2574|352100.0| NEAR BAY| 8.288135593220339|0.12951601908657123| 2.8022598870056497| 3.0|
 -122.25| 37.85| 52.0| 1274.0| 235.0| 558.0| 219.0| 5.6431|341300.0| NEAR BAY| 5.8173515981735155|0.18445839874411302| 2.547945205479452| 3.0|
 -122.25| 37.85| 52.0| 1627.0| 280.0| 565.0| 259.0| 3.8462|342200.0| NEAR BAY| 6.281853281853282| 0.1720958819913952| 2.1814671814671813| 3.0|
+---------+--------+------------------+-----------+--------------+----------+----------+-------------+--------+---------------+-------------------+-------------------+------------------------+--------------------+
only showing top 5 rows

Now we can use this numerical data in any Machine Learning algorithm. You can look at the mapping that this encoder has learned using the `labels` method: "<1H OCEAN" is mapped to 0, "INLAND" is mapped to 1, etc.

In [0]:
indexer.fit(renamedHousing).labels

Out[260]: ['<1H OCEAN', 'INLAND', 'NEAR OCEAN', 'NEAR BAY', 'ISLAND']

### One-hot encoding
Now, convert the label indices built in the last step into one-hot vectors. To do this, you can take advantage of the `OneHotEncoderEstimator` Estimator.

In [0]:
#// TODO: Replace <FILL IN> with appropriate code

from pyspark.ml.feature import OneHotEncoder


encoder = OneHotEncoder(
    inputCol="indx_ocean_proximity",
    outputCol="oh_ocean_proximity")  

ohHousing = encoder.fit(idxHousing).transform(idxHousing)

ohHousing.show(5)

+---------+--------+------------------+-----------+--------------+----------+----------+-------------+--------+---------------+-------------------+-------------------+------------------------+--------------------+------------------+
longitude|latitude|housing_median_age|total_rooms|total_bedrooms|population|households|median_income| label|ocean_proximity|rooms_per_household| bedrooms_per_room|population_per_household|indx_ocean_proximity|oh_ocean_proximity|
+---------+--------+------------------+-----------+--------------+----------+----------+-------------+--------+---------------+-------------------+-------------------+------------------------+--------------------+------------------+
 -122.23| 37.88| 41.0| 880.0| 129.0| 322.0| 126.0| 8.3252|452600.0| NEAR BAY| 6.984126984126984|0.14659090909090908| 2.5555555555555554| 3.0| (4,[3],[1.0])|
 -122.22| 37.86| 21.0| 7099.0| 1106.0| 2401.0| 1138.0| 8.3014|358500.0| NEAR BAY| 6.238137082601054|0.15579659106916466| 2.109841827768014| 3.0| (4,[3],[1.0])|
 -122.24| 37.85| 52.0| 1467.0| 190.0| 496.0| 177.0| 7.2574|352100.0| NEAR BAY| 8.288135593220339|0.12951601908657123| 2.8022598870056497| 3.0| (4,[3],[1.0])|
 -122.25| 37.85| 52.0| 1274.0| 235.0| 558.0| 219.0| 5.6431|341300.0| NEAR BAY| 5.8173515981735155|0.18445839874411302| 2.547945205479452| 3.0| (4,[3],[1.0])|
 -122.25| 37.85| 52.0| 1627.0| 280.0| 565.0| 259.0| 3.8462|342200.0| NEAR BAY| 6.281853281853282| 0.1720958819913952| 2.1814671814671813| 3.0| (4,[3],[1.0])|
+---------+--------+------------------+-----------+--------------+----------+----------+-------------+--------+---------------+-------------------+-------------------+------------------------+--------------------+------------------+
only showing top 5 rows

---
# 4. Pipeline
As you can see, there are many data transformation steps that need to be executed in the right order. For example, you called the `Imputer`, `VectorAssembler`, and `StandardScaler` from left to right. However, we can use the `Pipeline` class to define a sequence of Transformers/Estimators, and run them in order. A `Pipeline` is an `Estimator`, thus, after a Pipeline's `fit()` method runs, it produces a `PipelineModel`, which is a `Transformer`.

Now, let's create a pipeline called `numPipeline` to call the numerical transformers you built above (`imputer`, `va`, and `scaler`) in the right order from left to right, as well as a pipeline called `catPipeline` to call the categorical transformers (`indexer` and `encoder`). Then, put these two pipelines `numPipeline` and `catPipeline` into one pipeline.

In [0]:
#// TODO: Replace <FILL IN> with appropriate code#

#import org.apache.spark.ml.{Pipeline, PipelineModel, PipelineStage}
from pyspark.ml import Pipeline

numPipeline = Pipeline(stages=[imputer, assembler, scaler])
catPipeline = Pipeline(stages=[indexer, encoder])

pipeline = Pipeline(stages = [numPipeline, catPipeline])
newHousing = pipeline.fit(renamedHousing).transform(renamedHousing)

newHousing.show(5)

+---------+--------+------------------+-----------+--------------+----------+----------+-------------+--------+---------------+-------------------+-------------------+------------------------+--------------------+--------------------+--------------------+------------------+
longitude|latitude|housing_median_age|total_rooms|total_bedrooms|population|households|median_income| label|ocean_proximity|rooms_per_household| bedrooms_per_room|population_per_household| features| scaled_features|indx_ocean_proximity|oh_ocean_proximity|
+---------+--------+------------------+-----------+--------------+----------+----------+-------------+--------+---------------+-------------------+-------------------+------------------------+--------------------+--------------------+--------------------+------------------+
 -122.23| 37.88| 41.0| 880.0| 129.0| 322.0| 126.0| 8.3252|452600.0| NEAR BAY| 6.984126984126984|0.14659090909090908| 2.5555555555555554|[41.0,880.0,129.0...|[0.98211886567476...| 3.0| (4,[3],[1.0])|
 -122.22| 37.86| 21.0| 7099.0| 1106.0| 2401.0| 1138.0| 8.3014|358500.0| NEAR BAY| 6.238137082601054|0.15579659106916466| 2.109841827768014|[21.0,7099.0,1106...|[-0.6070042082805...| 3.0| (4,[3],[1.0])|
 -122.24| 37.85| 52.0| 1467.0| 190.0| 496.0| 177.0| 7.2574|352100.0| NEAR BAY| 8.288135593220339|0.12951601908657123| 2.8022598870056497|[52.0,1467.0,190....|[1.85613655635016...| 3.0| (4,[3],[1.0])|
 -122.25| 37.85| 52.0| 1274.0| 235.0| 558.0| 219.0| 5.6431|341300.0| NEAR BAY| 5.8173515981735155|0.18445839874411302| 2.547945205479452|[52.0,1274.0,235....|[1.85613655635016...| 3.0| (4,[3],[1.0])|
 -122.25| 37.85| 52.0| 1627.0| 280.0| 565.0| 259.0| 3.8462|342200.0| NEAR BAY| 6.281853281853282| 0.1720958819913952| 2.1814671814671813|[52.0,1627.0,280....|[1.85613655635016...| 3.0| (4,[3],[1.0])|
+---------+--------+------------------+-----------+--------------+----------+----------+-------------+--------+---------------+-------------------+-------------------+------------------------+--------------------+--------------------+--------------------+------------------+
only showing top 5 rows

Now, use `VectorAssembler` to put all attributes of the final dataset `newHousing` into a big vector, and call the new column `features_new`.

In [0]:
#// TODO: Replace <FILL IN> with appropriate code
va2 = VectorAssembler(
    inputCols=["scaled_features", "oh_ocean_proximity"],
    outputCol="features_new")

dataset = va2.transform(newHousing).select("features_new", "label")

dataset.show(5, truncate = False)

+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+--------+
features_new |label |
+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+--------+
[0.9821188656747666,-0.804799599801809,-0.9724529206674336,-0.9744049915469923,-0.977009185045236,2.3447089561176147,0.6285442264151613,-1.149894671563979,-0.049595334664572825,0.0,0.0,0.0,1.0] |452600.0|
[-0.6070042082805048,2.0458405373571247,1.3571105530973857,0.8614179998296625,1.6699205725917976,2.3321814648403056,0.32703343493534437,-0.990349761433348,-0.09250998908805363,0.0,0.0,0.0,1.0] |358500.0|
[1.8561365563501657,-0.5357329073251578,-0.8270042288561193,-0.8207574684591915,-0.8436164798678757,1.782656217213844,1.155592470660819,-1.4458218823526268,-0.025841901757064786,0.0,0.0,0.0,1.0] |352100.0|
[1.8561365563501657,-0.6241994689060669,-0.7197060135854777,-0.7660095004623889,-0.7337636638394612,0.9329449075937323,0.15696227955717026,-0.4936081250741483,-0.050328081993623826,0.0,0.0,0.0,1.0]|341300.0|
[1.8561365563501657,-0.4623927526466838,-0.612407798314836,-0.7598282782692015,-0.6291419342885903,-0.012880683843038048,0.344702448698221,-0.7078645624526793,-0.08561368522679019,0.0,0.0,0.0,1.0] |342200.0|
+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+--------+
only showing top 5 rows

In [0]:
dataset = dataset.withColumnRenamed('features_new', 'features')
dataset.show(5)

+--------------------+--------+
 features| label|
+--------------------+--------+
[0.98211886567476...|452600.0|
[-0.6070042082805...|358500.0|
[1.85613655635016...|352100.0|
[1.85613655635016...|341300.0|
[1.85613655635016...|342200.0|
+--------------------+--------+
only showing top 5 rows

---
# 5. Make a model
Here we going to make four different regression models:
* Linear regression model
* Decission tree regression
* Random forest regression
* Gradient-booster forest regression

But, before giving the data to train a Machine Learning model, let's first split the data into training dataset (`trainSet`) with 80% of the whole data, and test dataset (`testSet`) with 20% of it.

In [0]:
#// TODO: Replace <FILL IN> with appropriate code

trainSet, testSet = dataset.randomSplit([0.8, 0.2])
trainSet.show(5)
testSet.show(5)

+--------------------+--------+
 features| label|
+--------------------+--------+
[-2.1961272822357...| 55000.0|
[-2.1961272822357...|141700.0|
[-2.1961272822357...|191300.0|
[-2.1961272822357...|189200.0|
[-2.1166711285380...| 47500.0|
+--------------------+--------+
only showing top 5 rows

+--------------------+--------+
 features| label|
+--------------------+--------+
[-2.1166711285380...| 60000.0|
[-2.1166711285380...|310000.0|
[-2.1166711285380...|450000.0|
[-2.1166711285380...|100000.0|
[-2.1166711285380...|187500.0|
+--------------------+--------+
only showing top 5 rows

## 5.1. Linear regression model
Now, train a Linear Regression model using the `LinearRegression` class. Then, print the coefficients and intercept of the model, as well as the summary of the model over the training set by calling the `summary` method.

In [0]:
#// TODO: Replace <FILL IN> with appropriate code
from pyspark.ml.regression import LinearRegression

#// train the model
lr = LinearRegression(maxIter=10, regParam=0.3, elasticNetParam=0.8)
lrModel = lr.fit(trainSet)
trainingSummary = lrModel.summary

print(trainingSummary)
print("Coefficients: %s" % str(lrModel.coefficients))
print("Intercept: %s" % str(lrModel.intercept))
print("RMSE: %f" % trainingSummary.rootMeanSquaredError)

<pyspark.ml.regression.LinearRegressionTrainingSummary object at 0x7f11ba448c50>
Coefficients: [15101.648099157888,9556.528490592444,14245.361094180185,-35691.05152009123,17782.90806593698,82197.29196642565,2460.6320568692067,17893.873436032572,-502.9265562122965,15177.285716510882,-48043.29510216085,29012.879216120287,23367.68105291247]
Intercept: 208968.37229166954
RMSE: 69726.451869

Now, use `RegressionEvaluator` to measure the root-mean-square-erroe (RMSE) of the model on the test dataset.

In [0]:
#// TODO: Replace <FILL IN> with appropriate code

from pyspark.ml.evaluation import RegressionEvaluator

#// make predictions on the test data
predictions = lrModel.transform(testSet)
predictions.select("prediction", "label", "features").show(5)

#// select (prediction, true label) and compute test error.
evaluator = RegressionEvaluator(
    labelCol="label", predictionCol="prediction", metricName="rmse")
rmse = evaluator.evaluate(predictions)
print("Root Mean Squared Error (RMSE) on test data = %g" % rmse)

+------------------+--------+--------------------+
 prediction| label| features|
+------------------+--------+--------------------+
 153313.1219678202| 60000.0|[-2.1166711285380...|
317956.09560741426|310000.0|[-2.1166711285380...|
136644.20351597076|450000.0|[-2.1166711285380...|
122881.13020066771|100000.0|[-2.1166711285380...|
239829.81179922176|187500.0|[-2.1166711285380...|
+------------------+--------+--------------------+
only showing top 5 rows

Root Mean Squared Error (RMSE) on test data = 68076.4

## 5.2. Decision tree regression
Repeat what you have done on Regression Model to build a Decision Tree model. Use the `DecisionTreeRegressor` to make a model and then measure its RMSE on the test dataset.

In [0]:
#// TODO: Replace <FILL IN> with appropriate code

from pyspark.ml.regression import DecisionTreeRegressor

dt = DecisionTreeRegressor(featuresCol="features")

#// train the model
dtModel = dt.fit(trainSet)

#// make predictions on the test data
predictions = dtModel.transform(testSet)
predictions.select("prediction", "label", "features").show(5)

#// select (prediction, true label) and compute test error
evaluator = RegressionEvaluator(
    labelCol="label", predictionCol="prediction", metricName="rmse")
rmse = evaluator.evaluate(predictions)
print("Root Mean Squared Error (RMSE) on test data = %g" % rmse)

+------------------+--------+--------------------+
 prediction| label| features|
+------------------+--------+--------------------+
226526.65155807364| 60000.0|[-2.1166711285380...|
 324664.0869565217|310000.0|[-2.1166711285380...|
 186511.2992481203|450000.0|[-2.1166711285380...|
161789.92949640288|100000.0|[-2.1166711285380...|
 264362.1145902731|187500.0|[-2.1166711285380...|
+------------------+--------+--------------------+
only showing top 5 rows

Root Mean Squared Error (RMSE) on test data = 67536.1

## 5.3. Random forest regression
Let's try the test error on a Random Forest Model. Youcan use the `RandomForestRegressor` to make a Random Forest model.

In [0]:
#// TODO: Replace <FILL IN> with appropriate code

from pyspark.ml.regression import RandomForestRegressor

rf = RandomForestRegressor(featuresCol="features")
rfModel = rf.fit(trainSet)

#// make predictions on the test data
predictions = rfModel.transform(testSet)
predictions.select("prediction", "label", "features").show(5)

# #// select (prediction, true label) and compute test error
evaluator = RegressionEvaluator(
    labelCol="label", predictionCol="prediction", metricName="rmse")
rmse = evaluator.evaluate(predictions)
print("Root Mean Squared Error (RMSE) on test data = %g" % rmse)

+------------------+--------+--------------------+
 prediction| label| features|
+------------------+--------+--------------------+
 204957.0605010267| 60000.0|[-2.1166711285380...|
 356396.566692623|310000.0|[-2.1166711285380...|
189847.99644553248|450000.0|[-2.1166711285380...|
152099.36242271742|100000.0|[-2.1166711285380...|
 266218.7211857391|187500.0|[-2.1166711285380...|
+------------------+--------+--------------------+
only showing top 5 rows

Root Mean Squared Error (RMSE) on test data = 65404

## 5.4. Gradient-boosted tree regression
Fianlly, we want to build a Gradient-boosted Tree Regression model and test the RMSE of the test data. Use the `GBTRegressor` to build the model.

In [0]:
#// TODO: Replace <FILL IN> with appropriate code

from pyspark.ml.regression import GBTRegressor

gb = GBTRegressor(featuresCol="features", maxIter=10)

#// train the model
gbModel = gb.fit(trainSet)
#// make predictions on the test data
predictions = gbModel.transform(testSet)
predictions.select("prediction", "label", "features").show(5)

#// select (prediction, true label) and compute test error
evaluator = RegressionEvaluator(
    labelCol="label", predictionCol="prediction", metricName="rmse")
rmse = evaluator.evaluate(predictions)
print("Root Mean Squared Error (RMSE) on test data = %g" % rmse)

+-----------------+--------+--------------------+
 prediction| label| features|
+-----------------+--------+--------------------+
215805.0722987522| 60000.0|[-2.1166711285380...|
266178.8161371465|310000.0|[-2.1166711285380...|
 221271.360727282|450000.0|[-2.1166711285380...|
172563.4943279693|100000.0|[-2.1166711285380...|
243759.5080688684|187500.0|[-2.1166711285380...|
+-----------------+--------+--------------------+
only showing top 5 rows

Root Mean Squared Error (RMSE) on test data = 62112.4

---
# 6. Hyperparameter tuning
An important task in Machie Learning is model selection, or using data to find the best model or parameters for a given task. This is also called tuning. Tuning may be done for individual Estimators such as LinearRegression, or for entire Pipelines which include multiple algorithms, featurization, and other steps. Users can tune an entire Pipeline at once, rather than tuning each element in the Pipeline separately. MLlib supports model selection tools, such as `CrossValidator`. These tools require the following items:
* Estimator: algorithm or Pipeline to tune (`setEstimator`)
* Set of ParamMaps: parameters to choose from, sometimes called a "parameter grid" to search over (`setEstimatorParamMaps`)
* Evaluator: metric to measure how well a fitted Model does on held-out test data (`setEvaluator`)

`CrossValidator` begins by splitting the dataset into a set of folds, which are used as separate training and test datasets. For example with `k=3` folds, `CrossValidator` will generate 3 (training, test) dataset pairs, each of which uses 2/3 of the data for training and 1/3 for testing. To evaluate a particular `ParamMap`, `CrossValidator` computes the average evaluation metric for the 3 Models produced by fitting the Estimator on the 3 different (training, test) dataset pairs. After identifying the best `ParamMap`, `CrossValidator` finally re-fits the Estimator using the best ParamMap and the entire dataset.

Below, use the `CrossValidator` to select the best Random Forest model. To do so, you need to define a grid of parameters. Let's say we want to do the search among the different number of trees (1, 5, and 10), and different tree depth (5, 10, and 15).

In [0]:
#// TODO: Replace <FILL IN> with appropriate code

# import org.apache.spark.ml.tuning.ParamGridBuilder
# import org.apache.spark.ml.evaluation.RegressionEvaluator
# import org.apache.spark.ml.tuning.CrossValidator
from pyspark.ml.tuning import CrossValidator, ParamGridBuilder
from pyspark.ml.evaluation import RegressionEvaluator

#pipeline = Pipeline(stages=[rf])

#val paramGrid = new ParamGridBuilder().<FILL IN>
paramGrid = ParamGridBuilder() \
    .addGrid(rf.numTrees, [1, 5, 10]) \
    .addGrid(rf.maxDepth, [5, 10, 15]) \
    .build()

evaluator = RegressionEvaluator(
    labelCol="label", predictionCol="prediction", metricName="rmse")

crossval = CrossValidator(estimator=rf,
                          estimatorParamMaps=paramGrid,
                          evaluator=RegressionEvaluator(),
                          numFolds=3)


cvModel = crossval.fit(trainSet)

In [0]:
prediction = cvModel.transform(testSet)
predictions.select("prediction", "label", "features").show(5)

rmse = evaluator.evaluate(predictions)
print("Root Mean Squared Error (RMSE) on test data = %g" % rmse)

+-----------------+--------+--------------------+
 prediction| label| features|
+-----------------+--------+--------------------+
215805.0722987522| 60000.0|[-2.1166711285380...|
266178.8161371465|310000.0|[-2.1166711285380...|
 221271.360727282|450000.0|[-2.1166711285380...|
172563.4943279693|100000.0|[-2.1166711285380...|
243759.5080688684|187500.0|[-2.1166711285380...|
+-----------------+--------+--------------------+
only showing top 5 rows

Root Mean Squared Error (RMSE) on test data = 62112.4

---
# 7. An End-to-End Classification Test
As the last step, you are given a dataset called `data/ccdefault.csv`. The dataset represents default of credit card clients. It has 30,000 cases and 24 different attributes. More details about the dataset is available at `data/ccdefault.txt`. In this task you should make three models, compare their results and conclude the ideal solution. Here are the suggested steps:
1. Load the data.
2. Carry out some exploratory analyses (e.g., how various features and the target variable are distributed).
3. Train a model to predict the target variable (risk of `default`).
  - Employ three different models (logistic regression, decision tree, and random forest).
  - Compare the models' performances (e.g., AUC).
  - Defend your choice of best model (e.g., what are the strength and weaknesses of each of these models?).
4. What more would you do with this data? Anything to help you devise a better solution?

In [0]:
#Import data
credit = spark.read.format("csv").load("dbfs:/FileStore/shared_uploads/ravib@kth.se/ccdefault.csv",header=True, inferSchema= True)
credit.show(5)

+---+---------+---+---------+--------+---+-----+-----+-----+-----+-----+-----+---------+---------+---------+---------+---------+---------+--------+--------+--------+--------+--------+--------+-------+
 ID|LIMIT_BAL|SEX|EDUCATION|MARRIAGE|AGE|PAY_0|PAY_2|PAY_3|PAY_4|PAY_5|PAY_6|BILL_AMT1|BILL_AMT2|BILL_AMT3|BILL_AMT4|BILL_AMT5|BILL_AMT6|PAY_AMT1|PAY_AMT2|PAY_AMT3|PAY_AMT4|PAY_AMT5|PAY_AMT6|DEFAULT|
+---+---------+---+---------+--------+---+-----+-----+-----+-----+-----+-----+---------+---------+---------+---------+---------+---------+--------+--------+--------+--------+--------+--------+-------+
 1| 20000| 2| 2| 1| 24| 2| 2| -1| -1| -2| -2| 3913| 3102| 689| 0| 0| 0| 0| 689| 0| 0| 0| 0| 1|
 2| 120000| 2| 2| 2| 26| -1| 2| 0| 0| 0| 2| 2682| 1725| 2682| 3272| 3455| 3261| 0| 1000| 1000| 1000| 0| 2000| 1|
 3| 90000| 2| 2| 2| 34| 0| 0| 0| 0| 0| 0| 29239| 14027| 13559| 14331| 14948| 15549| 1518| 1500| 1000| 1000| 1000| 5000| 0|
 4| 50000| 2| 2| 1| 37| 0| 0| 0| 0| 0| 0| 46990| 48233| 49291| 28314| 28959| 29547| 2000| 2019| 1200| 1100| 1069| 1000| 0|
 5| 50000| 1| 2| 1| 57| -1| 0| -1| 0| 0| 0| 8617| 5670| 35835| 20940| 19146| 19131| 2000| 36681| 10000| 9000| 689| 679| 0|
+---+---------+---+---------+--------+---+-----+-----+-----+-----+-----+-----+---------+---------+---------+---------+---------+---------+--------+--------+--------+--------+--------+--------+-------+
only showing top 5 rows

---
Data Analysis

In [0]:
print('Schema')
credit.printSchema()

print('Number of Records')
credit.count()

Schema
root
-- ID: integer (nullable = true)
-- LIMIT_BAL: integer (nullable = true)
-- SEX: integer (nullable = true)
-- EDUCATION: integer (nullable = true)
-- MARRIAGE: integer (nullable = true)
-- AGE: integer (nullable = true)
-- PAY_0: integer (nullable = true)
-- PAY_2: integer (nullable = true)
-- PAY_3: integer (nullable = true)
-- PAY_4: integer (nullable = true)
-- PAY_5: integer (nullable = true)
-- PAY_6: integer (nullable = true)
-- BILL_AMT1: integer (nullable = true)
-- BILL_AMT2: integer (nullable = true)
-- BILL_AMT3: integer (nullable = true)
-- BILL_AMT4: integer (nullable = true)
-- BILL_AMT5: integer (nullable = true)
-- BILL_AMT6: integer (nullable = true)
-- PAY_AMT1: integer (nullable = true)
-- PAY_AMT2: integer (nullable = true)
-- PAY_AMT3: integer (nullable = true)
-- PAY_AMT4: integer (nullable = true)
-- PAY_AMT5: integer (nullable = true)
-- PAY_AMT6: integer (nullable = true)
-- DEFAULT: integer (nullable = true)

Number of Records
Out[196]: 30000

In [0]:
credit.agg({'LIMIT_BAL': 'min'}).show()
credit.agg({'LIMIT_BAL': 'max'}).show()
credit.agg({'LIMIT_BAL': 'avg'}).show()

credit.agg({'AGE': 'min'}).show()
credit.agg({'AGE': 'max'}).show()
credit.agg({'AGE': 'avg'}).show()

credit.agg({'PAY_3': 'min'}).show()
credit.agg({'PAY_3': 'max'}).show()
credit.agg({'PAY_3': 'avg'}).show()

credit.agg({'BILL_AMT3': 'min'}).show()
credit.agg({'BILL_AMT3': 'max'}).show()
credit.agg({'BILL_AMT3': 'avg'}).show()

credit.agg({'PAY_AMT3': 'min'}).show()
credit.agg({'PAY_AMT3': 'max'}).show()
credit.agg({'PAY_AMT3': 'avg'}).show()

+--------------+
min(LIMIT_BAL)|
+--------------+
 10000|
+--------------+

+--------------+
max(LIMIT_BAL)|
+--------------+
 1000000|
+--------------+

+------------------+
 avg(LIMIT_BAL)|
+------------------+
167484.32266666667|
+------------------+

+--------+
min(AGE)|
+--------+
 21|
+--------+

+--------+
max(AGE)|
+--------+
 79|
+--------+

+--------+
avg(AGE)|
+--------+
 35.4855|
+--------+

+----------+
min(PAY_3)|
+----------+
 -2|
+----------+

+----------+
max(PAY_3)|
+----------+
 8|
+----------+

+----------+
avg(PAY_3)|
+----------+
 -0.1662|
+----------+

+--------------+
min(BILL_AMT3)|
+--------------+
 -157264|
+--------------+

+--------------+
max(BILL_AMT3)|
+--------------+
 1664089|
+--------------+

+--------------+
avg(BILL_AMT3)|
+--------------+
 47013.1548|
+--------------+

+-------------+
min(PAY_AMT3)|
+-------------+
 0|
+-------------+

+-------------+
max(PAY_AMT3)|
+-------------+
 896040|
+-------------+

+-------------+
avg(PAY_AMT3)|
+-------------+
 5225.6815|
+-------------+

+------------+
min(DEFAULT)|
+------------+
 0|
+------------+

+------------+
max(DEFAULT)|
+------------+
 1|
+------------+

+------------+
avg(DEFAULT)|
+------------+
 0.2212|
+------------+

In [0]:
credit.groupBy('DEFAULT').count().sort(col("count").desc()).show()

+-------+-----+
DEFAULT|count|
+-------+-----+
 0|23364|
 1| 6636|
+-------+-----+

In [0]:
#Correlation
from pyspark.ml.feature import VectorAssembler
from pyspark.ml.stat import Correlation

assembler = VectorAssembler(
    inputCols=["LIMIT_BAL", "SEX", "EDUCATION", "MARRIAGE", "AGE", "DEFAULT"],
    outputCol="creditAttrs")

output = assembler.transform(credit)

r1 = Correlation.corr(output, "creditAttrs").head()
print(r1)
print("Pearson correlation matrix:\n" + str(r1[0]))

Row(pearson(creditAttrs)=DenseMatrix(6, 6, [1.0, 0.0248, -0.2192, -0.1081, 0.1447, -0.1535, 0.0248, 1.0, ..., 1.0, 0.0139, -0.1535, -0.04, 0.028, -0.0243, 0.0139, 1.0], False))
Pearson correlation matrix:
DenseMatrix([[ 1. , 0.02475524, -0.2191607 , -0.10813941, 0.1447128 ,
 -0.15351988],
 [ 0.02475524, 1. , 0.01423194, -0.03138884, -0.09087365,
 -0.03996058],
 [-0.2191607 , 0.01423194, 1. , -0.14346434, 0.17506066,
 0.02800608],
 [-0.10813941, -0.03138884, -0.14346434, 1. , -0.41416992,
 -0.02433922],
 [ 0.1447128 , -0.09087365, 0.17506066, -0.41416992, 1. ,
 0.01388983],
 [-0.15351988, -0.03996058, 0.02800608, -0.02433922, 0.01388983,
 1. ]])

---
Data Analysis

In [0]:
#Rename "deafult" to "label"
renamedCredit = credit.withColumnRenamed("DEFAULT", "label")
renamedCredit.show(2)

+---+---------+---+---------+--------+---+-----+-----+-----+-----+-----+-----+---------+---------+---------+---------+---------+---------+--------+--------+--------+--------+--------+--------+-----+
 ID|LIMIT_BAL|SEX|EDUCATION|MARRIAGE|AGE|PAY_0|PAY_2|PAY_3|PAY_4|PAY_5|PAY_6|BILL_AMT1|BILL_AMT2|BILL_AMT3|BILL_AMT4|BILL_AMT5|BILL_AMT6|PAY_AMT1|PAY_AMT2|PAY_AMT3|PAY_AMT4|PAY_AMT5|PAY_AMT6|label|
+---+---------+---+---------+--------+---+-----+-----+-----+-----+-----+-----+---------+---------+---------+---------+---------+---------+--------+--------+--------+--------+--------+--------+-----+
 1| 20000| 2| 2| 1| 24| 2| 2| -1| -1| -2| -2| 3913| 3102| 689| 0| 0| 0| 0| 689| 0| 0| 0| 0| 1|
 2| 120000| 2| 2| 2| 26| -1| 2| 0| 0| 0| 2| 2682| 1725| 2682| 3272| 3455| 3261| 0| 1000| 1000| 1000| 0| 2000| 1|
+---+---------+---+---------+--------+---+-----+-----+-----+-----+-----+-----+---------+---------+---------+---------+---------+---------+--------+--------+--------+--------+--------+--------+-----+
only showing top 2 rows

In [0]:
#Remove "ID"
renamedCredit = renamedCredit.drop('ID')

In [0]:
#Split into categorical and numerical. Remove labels
colNum = renamedCredit.schema.names

colNum.remove('label')
colNum.remove('SEX')
colNum.remove('EDUCATION')
colCat = ['SEX','EDUCATION','MARRIAGE']
print(colCat)
print(colNum)

['SEX', 'EDUCATION', 'MARRIAGE']
['LIMIT_BAL', 'MARRIAGE', 'AGE', 'PAY_0', 'PAY_2', 'PAY_3', 'PAY_4', 'PAY_5', 'PAY_6', 'BILL_AMT1', 'BILL_AMT2', 'BILL_AMT3', 'BILL_AMT4', 'BILL_AMT5', 'BILL_AMT6', 'PAY_AMT1', 'PAY_AMT2', 'PAY_AMT3', 'PAY_AMT4', 'PAY_AMT5', 'PAY_AMT6']

In [0]:
#Check for missing data
renamedCredit.select([count(when(col(c).isNull(), c)).alias(c) for c in renamedCredit.columns]).show()

#No missing data

+---------+---+---------+--------+---+-----+-----+-----+-----+-----+-----+---------+---------+---------+---------+---------+---------+--------+--------+--------+--------+--------+--------+-----+
LIMIT_BAL|SEX|EDUCATION|MARRIAGE|AGE|PAY_0|PAY_2|PAY_3|PAY_4|PAY_5|PAY_6|BILL_AMT1|BILL_AMT2|BILL_AMT3|BILL_AMT4|BILL_AMT5|BILL_AMT6|PAY_AMT1|PAY_AMT2|PAY_AMT3|PAY_AMT4|PAY_AMT5|PAY_AMT6|label|
+---------+---+---------+--------+---+-----+-----+-----+-----+-----+-----+---------+---------+---------+---------+---------+---------+--------+--------+--------+--------+--------+--------+-----+
 0| 0| 0| 0| 0| 0| 0| 0| 0| 0| 0| 0| 0| 0| 0| 0| 0| 0| 0| 0| 0| 0| 0| 0|
+---------+---+---------+--------+---+-----+-----+-----+-----+-----+-----+---------+---------+---------+---------+---------+---------+--------+--------+--------+--------+--------+--------+-----+

In [0]:
#Scaling
from pyspark.ml.feature import VectorAssembler, StandardScaler

assembler_cat = VectorAssembler(
    inputCols=colNum,
    outputCol="numerical_attributess")

featuredCredit = assembler.transform(renamedCredit)


scaler_cat = StandardScaler(
  inputCol="numerical_attributess",
  outputCol="scaled_num_attributes",
  withStd = True,
  withMean = True)
 
scaledCredit = scaler.fit(featuredCredit).transform(featuredCredit)

scaledCredit.select("numerical_attributess", "scaled_num_attributes").show(5, truncate=False)

+------------------------------------------------------------------------------------------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
numerical_attributess |scaled_num_attributes |
+------------------------------------------------------------------------------------------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
[20000.0,1.0,24.0,2.0,2.0,-1.0,-1.0,-2.0,-2.0,3913.0,3102.0,689.0,0.0,0.0,0.0,0.0,689.0,0.0,0.0,0.0,0.0] |[-1.1367012005089403,-1.0572774085277983,-1.2459990812617805,1.7945339544019396,1.7823184657418991,-0.6966518451846738,-0.6665876213480703,-1.5300205339519923,-1.4860159900387195,-0.6424903644804023,-0.6473884403600334,-0.6679821771856058,-0.67248606025838,-0.6630474806912302,-0.6527133393110748,-0.34193591963342385,-0.22708185101951875,-0.2967963273482676,-0.3080574273159171,-0.31413088166475134,-0.2933771681829596] |
[120000.0,2.0,26.0,-1.0,2.0,0.0,0.0,0.0,2.0,2682.0,1725.0,2682.0,3272.0,3455.0,3261.0,0.0,1000.0,1000.0,1000.0,0.0,2000.0] |[-0.3659744005642241,0.8585429741633135,-1.0290300191814483,-0.8749765643692307,1.7823184657418991,0.1388624810142633,0.18874294496681934,0.23491260014881957,1.9922823072021227,-0.6592077626391072,-0.6667354557500351,-0.6392436392699414,-0.6216255793683517,-0.6062191643624951,-0.597956410949955,-0.34193591963342385,-0.2135840970469851,-0.2400006103976141,-0.24422557468968872,-0.31413088166475134,-0.18087519375848451] |
[90000.0,2.0,34.0,0.0,0.0,0.0,0.0,0.0,0.0,29239.0,14027.0,13559.0,14331.0,14948.0,15549.0,1518.0,1500.0,1000.0,1000.0,1000.0,5000.0]|[-0.597192440547639,0.8585429741633135,-0.1611537708601192,0.014860275221159448,0.11173424152941251,0.1388624810142633,0.18874294496681934,0.23491260014881957,0.2531331585817016,-0.2985546814849851,-0.4938908781699325,-0.4824001485982493,-0.44972274985648625,-0.41718071586288513,-0.3916230360031781,-0.25028740752989925,-0.19188352795931046,-0.2400006103976141,-0.24422557468968872,-0.24867859781857737,-0.012122232121771874] |
[50000.0,1.0,37.0,0.0,0.0,0.0,0.0,0.0,0.0,46990.0,48233.0,49291.0,28314.0,28959.0,29547.0,2000.0,2019.0,1200.0,1100.0,1069.0,1000.0]|[-0.9054831605255255,-1.0572774085277983,0.1642998222603792,0.014860275221159448,0.11173424152941251,0.1388624810142633,0.18874294496681934,0.23491260014881957,0.2531331585817016,-0.05749007164325524,-0.01329246972374568,0.03284593108189235,-0.23236880600402093,-0.18672585621544444,-0.1565762763592179,-0.2211868918290436,-0.16935833724630417,-0.22864146700748342,-0.23784238942706587,-0.24416239023319136,-0.23712618097072205]|
[50000.0,1.0,57.0,-1.0,0.0,-1.0,0.0,0.0,0.0,8617.0,5670.0,35835.0,20940.0,19146.0,19131.0,2000.0,36681.0,10000.0,9000.0,689.0,679.0]|[-0.9054831605255255,-1.0572774085277983,2.3339904430637017,-0.8749765643692307,0.11173424152941251,-0.6966518451846738,0.18874294496681934,0.23491260014881957,0.2531331585817016,-0.5786084465774393,-0.6113077319202916,-0.1611860639974759,-0.3469914178875995,-0.34813143309934536,-0.33147605121644674,-0.2211868918290436,1.335011914187654,0.271160842158267,0.2664292463201387,-0.26903425809473747,-0.2551827478658503] |
+----------------------------------------------------------------------

In [0]:
#One Hot Encoding
from pyspark.ml.feature import OneHotEncoder

encoder_cat = OneHotEncoder(
    inputCols=['SEX', 'EDUCATION', 'MARRIAGE'],
    outputCols=['OH_SEX', 'OH_EDUCATION', 'OH_MARRIAGE'])  

ohCredit = encoder_cat.fit(scaledCredit).transform(scaledCredit)

ohCredit.show(5)

+---------+---+---------+--------+---+-----+-----+-----+-----+-----+-----+---------+---------+---------+---------+---------+---------+--------+--------+--------+--------+--------+--------+-----+---------------------+---------------------+-------------+-------------+-------------+
LIMIT_BAL|SEX|EDUCATION|MARRIAGE|AGE|PAY_0|PAY_2|PAY_3|PAY_4|PAY_5|PAY_6|BILL_AMT1|BILL_AMT2|BILL_AMT3|BILL_AMT4|BILL_AMT5|BILL_AMT6|PAY_AMT1|PAY_AMT2|PAY_AMT3|PAY_AMT4|PAY_AMT5|PAY_AMT6|label|numerical_attributess|scaled_num_attributes| OH_SEX| OH_EDUCATION| OH_MARRIAGE|
+---------+---+---------+--------+---+-----+-----+-----+-----+-----+-----+---------+---------+---------+---------+---------+---------+--------+--------+--------+--------+--------+--------+-----+---------------------+---------------------+-------------+-------------+-------------+
 20000| 2| 2| 1| 24| 2| 2| -1| -1| -2| -2| 3913| 3102| 689| 0| 0| 0| 0| 689| 0| 0| 0| 0| 1| [20000.0,1.0,24.0...| [-1.1367012005089...| (2,[],[])|(6,[2],[1.0])|(3,[1],[1.0])|
 120000| 2| 2| 2| 26| -1| 2| 0| 0| 0| 2| 2682| 1725| 2682| 3272| 3455| 3261| 0| 1000| 1000| 1000| 0| 2000| 1| [120000.0,2.0,26....| [-0.3659744005642...| (2,[],[])|(6,[2],[1.0])|(3,[2],[1.0])|
 90000| 2| 2| 2| 34| 0| 0| 0| 0| 0| 0| 29239| 14027| 13559| 14331| 14948| 15549| 1518| 1500| 1000| 1000| 1000| 5000| 0| [90000.0,2.0,34.0...| [-0.5971924405476...| (2,[],[])|(6,[2],[1.0])|(3,[2],[1.0])|
 50000| 2| 2| 1| 37| 0| 0| 0| 0| 0| 0| 46990| 48233| 49291| 28314| 28959| 29547| 2000| 2019| 1200| 1100| 1069| 1000| 0| [50000.0,1.0,37.0...| [-0.9054831605255...| (2,[],[])|(6,[2],[1.0])|(3,[1],[1.0])|
 50000| 1| 2| 1| 57| -1| 0| -1| 0| 0| 0| 8617| 5670| 35835| 20940| 19146| 19131| 2000| 36681| 10000| 9000| 689| 679| 0| [50000.0,1.0,57.0...| [-0.9054831605255...|(2,[1],[1.0])|(6,[2],[1.0])|(3,[1],[1.0])|
+---------+---+---------+--------+---+-----+-----+-----+-----+-----+-----+---------+---------+---------+---------+---------+---------+--------+--------+--------+--------+--------+--------+-----+---------------------+---------------------+-------------+-------------+-------------+
only showing top 5 rows

In [0]:
from pyspark.ml import Pipeline

numPipeline = Pipeline(stages=[assembler_cat, scaler_cat])
catPipeline = Pipeline(stages=[encoder_cat])

pipeline = Pipeline(stages = [numPipeline, catPipeline])
newCredit = pipeline.fit(renamedCredit).transform(renamedCredit)

newCredit = newCredit.drop('SEX', 'EDUCATION', 'MARRIAGE')

newCredit.show(5)




+---------+---+-----+-----+-----+-----+-----+-----+---------+---------+---------+---------+---------+---------+--------+--------+--------+--------+--------+--------+-----+---------------------+---------------------+-------------+-------------+-------------+
LIMIT_BAL|AGE|PAY_0|PAY_2|PAY_3|PAY_4|PAY_5|PAY_6|BILL_AMT1|BILL_AMT2|BILL_AMT3|BILL_AMT4|BILL_AMT5|BILL_AMT6|PAY_AMT1|PAY_AMT2|PAY_AMT3|PAY_AMT4|PAY_AMT5|PAY_AMT6|label|numerical_attributess|scaled_num_attributes| OH_SEX| OH_EDUCATION| OH_MARRIAGE|
+---------+---+-----+-----+-----+-----+-----+-----+---------+---------+---------+---------+---------+---------+--------+--------+--------+--------+--------+--------+-----+---------------------+---------------------+-------------+-------------+-------------+
 20000| 24| 2| 2| -1| -1| -2| -2| 3913| 3102| 689| 0| 0| 0| 0| 689| 0| 0| 0| 0| 1| [20000.0,1.0,24.0...| [-1.1367012005089...| (2,[],[])|(6,[2],[1.0])|(3,[1],[1.0])|
 120000| 26| -1| 2| 0| 0| 0| 2| 2682| 1725| 2682| 3272| 3455| 3261| 0| 1000| 1000| 1000| 0| 2000| 1| [120000.0,2.0,26....| [-0.3659744005642...| (2,[],[])|(6,[2],[1.0])|(3,[2],[1.0])|
 90000| 34| 0| 0| 0| 0| 0| 0| 29239| 14027| 13559| 14331| 14948| 15549| 1518| 1500| 1000| 1000| 1000| 5000| 0| [90000.0,2.0,34.0...| [-0.5971924405476...| (2,[],[])|(6,[2],[1.0])|(3,[2],[1.0])|
 50000| 37| 0| 0| 0| 0| 0| 0| 46990| 48233| 49291| 28314| 28959| 29547| 2000| 2019| 1200| 1100| 1069| 1000| 0| [50000.0,1.0,37.0...| [-0.9054831605255...| (2,[],[])|(6,[2],[1.0])|(3,[1],[1.0])|
 50000| 57| -1| 0| -1| 0| 0| 0| 8617| 5670| 35835| 20940| 19146| 19131| 2000| 36681| 10000| 9000| 689| 679| 0| [50000.0,1.0,57.0...| [-0.9054831605255...|(2,[1],[1.0])|(6,[2],[1.0])|(3,[1],[1.0])|
+---------+---+-----+-----+-----+-----+-----+-----+---------+---------+---------+---------+---------+---------+--------+--------+--------+--------+--------+--------+-----+---------------------+---------------------+-------------+-------------+-------------+
only showing top 5 rows

In [0]:
newCreditColumns = newCredit.schema.names
va2 = VectorAssembler(
    inputCols=["scaled_num_attributes", "OH_SEX", "OH_EDUCATION", "OH_MARRIAGE"],
    outputCol="features")

datasetcredit = va2.transform(newCredit).select("features", "label")

datasetcredit.show(5)

+--------------------+-----+
 features|label|
+--------------------+-----+
[-1.1367012005089...| 1|
[-0.3659744005642...| 1|
[-0.5971924405476...| 0|
[-0.9054831605255...| 0|
[-0.9054831605255...| 0|
+--------------------+-----+
only showing top 5 rows

In [0]:
trainSet, testSet = datasetcredit.randomSplit([0.8, 0.2])
trainSet.show(5)
testSet.show(5)

+--------------------+-----+
 features|label|
+--------------------+-----+
[-1.2137738805034...| 0|
[-1.2137738805034...| 0|
[-1.2137738805034...| 0|
[-1.2137738805034...| 0|
[-1.2137738805034...| 0|
+--------------------+-----+
only showing top 5 rows

+--------------------+-----+
 features|label|
+--------------------+-----+
[-1.2137738805034...| 0|
[-1.2137738805034...| 0|
[-1.2137738805034...| 1|
[-1.2137738805034...| 0|
[-1.2137738805034...| 1|
+--------------------+-----+
only showing top 5 rows

---
Model Performance

In [0]:
 #Logistic Regression

from pyspark.ml.classification import LogisticRegression
from pyspark.ml.evaluation import MulticlassClassificationEvaluator
from pyspark.mllib.evaluation import BinaryClassificationMetrics as metric

lr = LogisticRegression(maxIter=10)

# Fit the model
lrModel = lr.fit(trainSet)

# Print the coefficients and intercept for logistic regression
print("Coefficients: " + str(lrModel.coefficients))
print("Intercept: " + str(lrModel.intercept))

#// make predictions on the test data
predictions = lrModel.transform(testSet)
predictions.select("prediction", "label","probability").show(10,False)

evaluator = MulticlassClassificationEvaluator(
    labelCol="label", predictionCol="prediction", metricName="accuracy")
accuracy = evaluator.evaluate(predictions)
print("Test Error = %g " % (1.0 - accuracy))

results = predictions.select(['probability', 'label'])
 
## prepare score-label set
results_collect = results.collect()
results_list = [(float(i[0][0]), 1.0-float(i[1])) for i in results_collect]
scoreAndLabels = sc.parallelize(results_list)
 
metrics = metric(scoreAndLabels)
print("The ROC score is : ", metrics.areaUnderROC)

Coefficients: [-0.08083154461413497,-0.023108368649455157,0.05896844616723888,0.6892458898113276,0.09573061868477252,0.07148387579790726,0.054053682910255745,0.04515893692454439,-0.013589498434353035,-0.15073376855568738,-0.03195808099568721,-0.0328661099083384,0.018152049823681196,0.045927009982941455,0.054868716999919395,-0.18361511754674814,-0.15403174391452062,-0.05146983110966975,-0.062023266115104836,-0.07175943616936437,-0.07236630593879693,0.0,0.09329137119186086,-2.7663391214034205,0.0027580166269188717,-0.11078856027784405,-0.08926253561300085,-1.4854950038375367,-1.4581159678838738,-0.986239906964813,-0.05685344941258549,-0.1653676617725679]
Intercept: -1.32582991319728
+----------+-----+----------------------------------------+
prediction|label|probability |
+----------+-----+----------------------------------------+
0.0 |0 |[0.7422351123684169,0.25776488763158306]|
0.0 |0 |[0.7484446404175162,0.2515553595824838] |
0.0 |1 |[0.7491731172918287,0.2508268827081714] |
0.0 |0 |[0.749951370183354,0.25004862981664594] |
0.0 |1 |[0.8670258254820428,0.13297417451795715]|
0.0 |0 |[0.6831881678291138,0.3168118321708861] |
0.0 |1 |[0.5668846281404458,0.4331153718595541] |
1.0 |1 |[0.2095160624756365,0.7904839375243635] |
1.0 |0 |[0.29520140738138245,0.7047985926186175]|
0.0 |0 |[0.7069040893554648,0.29309591064453533]|
+----------+-----+----------------------------------------+
only showing top 10 rows

Test Error = 0.196432 
The ROC score is : 0.7018935976926791

In [0]:
#Decision Tree
from pyspark.ml.classification import DecisionTreeClassifier

dt = DecisionTreeClassifier(maxDepth=10)

# Train model.  This also runs the indexers.
dtModel = dt.fit(trainSet)
#dtmodel = pipeline.fit(trainSet)

# Make predictions.
predictions = dtModel.transform(testSet)

# Select example rows to display.
predictions.select("prediction", "label", "probability").show(5,False)

evaluator = MulticlassClassificationEvaluator(
    labelCol="label", predictionCol="prediction", metricName="accuracy")
accuracy = evaluator.evaluate(predictions)
print("Test Error = %g " % (1.0 - accuracy))

results = predictions.select(['probability', 'label'])
 
## prepare score-label set
results_collect = results.collect()
results_list = [(float(i[0][0]), 1.0-float(i[1])) for i in results_collect]
scoreAndLabels = sc.parallelize(results_list)
 
metrics = metric(scoreAndLabels)
print("The ROC score is : ", metrics.areaUnderROC)

+----------+-----+----------------------------------------+
prediction|label|probability |
+----------+-----+----------------------------------------+
0.0 |0 |[0.7366863905325444,0.26331360946745563]|
0.0 |0 |[0.8830845771144279,0.11691542288557213]|
0.0 |1 |[0.782608695652174,0.21739130434782608] |
0.0 |0 |[0.782608695652174,0.21739130434782608] |
0.0 |1 |[0.7800925925925926,0.2199074074074074] |
+----------+-----+----------------------------------------+
only showing top 5 rows

Test Error = 0.196765 
The ROC score is : 0.7227485833354509

In [0]:
#Random Forest
from pyspark.ml.classification import RandomForestClassifier
from pyspark.ml.evaluation import MulticlassClassificationEvaluator

rf = RandomForestClassifier(maxDepth=10)

# Train model.  This also runs the indexers.
rfModel = rf.fit(trainSet)

# Make predictions.
predictions = rfModel.transform(testSet)

# Select example rows to display.
predictions.select("prediction", "label", "probability").show(5,False)

evaluator = MulticlassClassificationEvaluator(
    labelCol="label", predictionCol="prediction", metricName="accuracy")
accuracy = evaluator.evaluate(predictions)
print("Test Error = %g " % (1.0 - accuracy))

results = predictions.select(['probability', 'label'])
 
## prepare score-label set
results_collect = results.collect()
results_list = [(float(i[0][0]), 1.0-float(i[1])) for i in results_collect]
scoreAndLabels = sc.parallelize(results_list)
 
metrics = metric(scoreAndLabels)
print("The ROC score is : ", metrics.areaUnderROC)

+----------+-----+----------------------------------------+
prediction|label|probability |
+----------+-----+----------------------------------------+
0.0 |0 |[0.8494074614312993,0.15059253856870075]|
0.0 |0 |[0.8631588161879363,0.13684118381206356]|
0.0 |1 |[0.8275595352733239,0.172440464726676] |
0.0 |0 |[0.8089990211309732,0.19100097886902678]|
0.0 |1 |[0.7067873239201038,0.29321267607989615]|
+----------+-----+----------------------------------------+
only showing top 5 rows

Test Error = 0.188428 
The ROC score is : 0.7627126585012579

---
Model Evaluations

---
First we will compare the model's performances. The highest AUC was obtained using Random Forest (0.76), followed by Decision Tree (0.72), followed by Logistic Regression (0.70). The higher the AUC, the better the model is at predicting 0s as 0s and 1s as 1s. Therefore the AUC tells us how good the model is at distinguishing between classes. Since Random Forest achieved the highest AUC, we can conclude that it is the best model at distinguishing between classes, followed by Decision Trees, and Logisitic Regression was the weakest model at distinguishing between classes. However, since the 3 AUC values obtained are quite similar to each other, we can say that the three models overall perform quite similarly when it comes to distinguishing between the classes and predicting the correct label, with Random Forest only slightly outperforming the others.

The random forest gives a good performance on both classification and regression tasks (although it is better at classification). Random forests are also good at avoiding overfitting during classification. They also deals with unequal classes well. However, sometimes a lot of trees are needed for random forests to be effective, which can be quite computationally expensive. In contrast, a decision tree does not require as much computational power, and also has a shorter training time. It is also very simple, and still has a decent performace. However, if the data needs to be slightly altered, then the structure of the tree would be greatly affected. Logistic regression is also fairly computationally inexpensive, and is simple and easy to interpret. However its performance is relatively poor compared to other methods, and it struggles with solving non-linear problems.

In order to derive a better solution we would perform some feature engineering. This would significantly increase some of these model's performances, especially logistic regression. We would also tune the hyperparameters of these models, as this would increase their performance aswell. Multiple different hyperparamter tuning techniques could be investigated. Finally we would try to understand the data better by gaining more knowledge into what it represents and why it is important.